### DISCUSSION TITLE
Best way to couple several apps

### DISCUSSION THREAD

**Discussion by aarograh** [idaholab/moose#21980] [Q&A General] [answered]
I'm working on coupling several apps together.  Up to this point We've only been coupling 2 apps, so the approach we took was to have one app's class inherit from the other to have full access to everything in both apps.  However, when adding a third app, this introduces a diamond inheritance that cannot be overcome (without modifying the parent apps, which isn't an option).  I'm aware people have compiled several apps together, so I'm sure the approach I've taken so far must not be the recommended approach.  Could someone point me toward something that would give an example of how I should actually do this?

**Comment by GiudGiud** [comment]
Hello

Could you please ask access to BlueCrab on the INL NCRC?
That will be a good example of how to combine applications. 

While the inner apps are restricted, BC is actually EAR-99 so we should be able to get you to see it very quickly.

Guillaume

**Reply by GiudGiud** [reply]
please let us know when you have made the request

**Reply by aarograh** [reply]
Ok, I'll do that.  Thanks.

**Comment by GiudGiud** [comment, marked as answer]
Cardinal is an open source example of this otherwise
https://github.com/neams-th-coe/cardinal/blob/devel/src/base/CardinalApp.C
you can see every app in there, with all the registering

**Reply by GiudGiud** [reply]
the Makefile also has the necessary mods, but it s more complicated because OpenMC and Nek arent moose apps

**Reply by aarograh** [reply]
Thanks for the info. I'll check that out.  I'll be working with a non-MOOSE app shortly too, so this might be a really helpful example

**Reply by aarograh** [reply]
@GiudGiud thanks again for this suggestion.  This has helped with sorting out some issues.  I do have a follow-up question.  I wanted to enable coupling between the 2 apps using both MultiApps and "full coupling".  Previously we accomplished "full coupling" by having `moleApp` inherit from `SamApp`.  I'm trying to move away from this approach since I"m guessing there's a better way of accomplishing that.

My current issue is that the "full coupling" inputs fail when trying to read the portions of the input related to SAM syntax.  However, if I try calling SAM's `SAM::associateSyntaxInner()` from `moleApp::setupOptions()` (SAM calls it from `SamApp::setupOptions()`, so I'm just mimicking SAM on this one) I just get a segfault.  Here's the backtrace on the segfault.  Any suggestions on the "proper" way to accomplish full coupling?  I tried looking around the MOOSE documentation but haven't found the answer yet.

```
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff786795e in SamApp::setSimulation (this=0x0, sim=0x555555d8a8d0) at /home/ag6/zoo/mole_sam/contrib/SAM/src/base/SamApp.C:47
47        _sim = sim;
(gdb) bt
#0  0x00007ffff786795e in SamApp::setSimulation (this=0x0, sim=0x555555d8a8d0) at /home/ag6/zoo/mole_sam/contrib/SAM/src/base/SamApp.C:47
#1  0x00007ffff7231d7f in SAMCreateMeshAction::act (this=0x555555b48760) at /home/ag6/zoo/mole_sam/contrib/SAM/src/actions/SAMCreateMeshAction.C:62
#2  0x00007ffff2e09a8b in Action::timedAct (this=0x555555b48760) at /home/ag6/zoo/mole_sam/moose/framework/src/actions/Action.C:92
#3  0x00007ffff2e0e402 in ActionWarehouse::executeActionsWithAction (this=0x55555582de20, task=...) at /home/ag6/zoo/mole_sam/moose/framework/src/actions/ActionWarehouse.C:384
#4  0x00007ffff2e0df2c in ActionWarehouse::executeAllActions (this=0x55555582de20) at /home/ag6/zoo/mole_sam/moose/framework/src/actions/ActionWarehouse.C:344
#5  0x00007ffff4310546 in MooseApp::runInputFile (this=0x55555582d840) at /home/ag6/zoo/mole_sam/moose/framework/src/base/MooseApp.C:1044
#6  0x00007ffff4313d4c in MooseApp::run (this=0x55555582d840) at /home/ag6/zoo/mole_sam/moose/framework/src/base/MooseApp.C:1403
#7  0x00005555555577c7 in main (argc=3, argv=0x7fffffffc2c8) at /home/ag6/zoo/mole_sam/src/main.C:33
```

**Reply by GiudGiud** [reply]
Full coupling with SAM is more complicated because SAM replaces some of the moose syntax, notably postprocessors syntax.

The SAMApp is null? (this = 0x0) A missing createApp call is likely. I think that's usually done in main.C. What kind of app do you create there?

**Reply by aarograh** [reply]
Yeah, as I was leaving for the day I noticed that in main.C and wondered if that might be part of the issues.  Here is what was in main.C already:
```
  std::shared_ptr<MooseApp> app = AppFactory::createAppShared("moleTestApp", argc, argv);
  app->run();
```
I add this line between the two above just on a whim:
```
  std::shared_ptr<MooseApp> sam_app = AppFactory::createAppShared("SamApp", argc, argv);)
```
However, that did not work.  I'm assuming the issue is that the `setSimulation` method on `SamApp` takes a `SAMSimulation *` but is being given an `FEProblem *` or something like that because `SamApp` is not driving the solve. 

I'm open to ideas you may have to resolve this, but at this point I may just switch these tests to be `MultiApps` tests instead, and not bother supporting the full coupling.  However, if you have any quick ideas, please let me know and I'll try them out.

**Reply by GiudGiud** [reply]
You would not create two apps, you would replace what you have with "moleTestApp" to "SamApp"

If there is indeed a cast from MooseApp to SamApp in SAMSimulation that could be the problem. I ll try to get my hands on SAM again and double check.

Are the simulation domains overlapping between SAM and Mole? Or are they coupled at boundary conditions?

**Reply by aarograh** [reply]
Using `SamApp` instead of `moleTestApp` won't work because then I don't have access to all the `mole` components I need.

The simulation domains are expected to be overlapping for now, yes.  However, I can foresee cases where that may not be the case (either they would be coupled at boundary conditions, or perhaps part of the domain would overlap but each code would have additional domain that the other does not need to model).

**Reply by GiudGiud** [reply]
ok then you need an app that sits on top of both.

Or if you have this liberty (eg you dont have to keep SAM and mole unmodified), you can register all the objects from the other app to one of the two apps. 

If they are overlapping there's a stronger case for full coupling for now

**Reply by aarograh** [reply]
If Mole is the "top" app, is it possible to register all of SAM's objects to Mole without modifying SAM?  In either case, where would I make the change?  To be honest, I thought that was already happening since I'm calling `SamApp::registerApps()` and `SamApp::registerAll(f, af, s)` from `moleApp::registerAll()`.

**Reply by GiudGiud** [reply]
yeah so I dont know if you can share this routine, maybe you can email it to me on my INL email
but basically it should be like (minus any mistake I made):
```
void
moleApp::registerAll(Factory & f, ActionFactory & af, Syntax & s)
{
  Registry::registerObjectsTo(f, {"moleApp"});
  Registry::registerActionsTo(af, {"moleApp"});

  ModulesApp::registerAll(f, af, s);

#ifdef SAM_ENABLED
  SamApp::registerAll(f, af, s);
  
  // SamApp::associateSyntax(s, af);
  // if you do that, you can run Sam input files directly, but might mess up regular input files
  // if you dont, you have to pass --app SamApp on the command line
#endif
}
```

**Reply by GiudGiud** [reply]
and then in registerApp
```
void
moleApp::registerApps()
{
  registerApp(moleApp);
#ifdef SAM_ENABLED
  SamApp::registerApps();
#endif
}
```

**Reply by aarograh** [reply]
Ok.  I had something very similar to that, but tweaked it to match what you have.  The version of `SamApp` I'm working with right now does not have `associateSyntax` as a method.  The closest I could come was `SAM::associateSyntaxInner(...)` (where `SAM` is a namespace, not the app).  I'm wondering if they changed it between the version I'm using and the current master (I'm currently working to update the version of moose/sam/griffin we use, so if that's the case, I'll get to the new version soon enough).

All that said, I still get the segfault after making those changes.

**Reply by GiudGiud** [reply]
the problem with associating syntax is that it wont be compatible across apps that modify syntax (eg SAM). So for putting everything in the same input file for full coupling, that s an issue. 

SamApp seems to be a nullptr. I d double confirm this by printing `this` in SamSimulation.
If so, how did this happen?
Likely there is a cast from the app to a SamApp pointer.
This cast should have been checked for null, it seems it s not. To make it not evaluate to null, you will have to make the app a SamApp, usually by having the app be a derived class of SamApp

**Reply by aarograh** [reply]
Yeah, ok.  I think what I'm hearing is "just worry about MultiApps for now".  I can revisit the full coupling approach later, perhaps with a separate app.  I don't want Mole inheriting from SAM because, as you said, that ends up modifying everything that Mole does.

I appreciate all the help.  Glad to learn a little more about this, and also learn that I wans't just being dumb and there really are some subtleties specifically with SAM.

**Reply by GiudGiud** [reply]
Yeah sorry. The way we deal with SAM in bluecrab is to pass `--app SamApp` on the command line, that way we create a SamApp.
We do that to run SAM input files only
I guess you can try this too, but dont expect full agreement on mole inputs.
